File: matlab/unit/demo_bj.m
%  Running this tests PEM estimation of Box-Jenkins model structure

clear; close all;  
global dm; if isempty(dm), clear global dm; dm=0; end
global dsp; if isempty(dsp), clear global dsp; dsp=1; end
global trans;
if dsp>1, clc; echo on; end
OPT.dsp = dsp;

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Specify Experiment Conditions                    
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
T   = 1;     % Sampling period in seconds
N   = 500;   % Number of samples
var = 1e-2;  % White Measurement Noise variance

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Specify True System                    
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

den     = real(poly([-0.1,-1,-0.001-0.2j,-0.001+0.2j,-0.5]));  % Cts time spec
num     = den(length(den));
[bq,aq] = c2dm(num,den,T,'zoh');  % Discrete time version
cq      = [1,-0.2];  
dq      = [1,-0.5];               % Measurement noise colouring

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Simulate a data record                    
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

t     = 0:1:N-1; 
Z.u   = sign(sin(5*pi*t/N));
%Z.u     = randn(1,N);
noise   = filter(cq,dq,sqrt(var)*randn(size(Z.u)));
Z.y     = filter(bq,aq,Z.u)+noise; 

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Specify Model Structure                    
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

Mq.A = length(aq)-1; 
Mq.B = length(bq)-2;  
Mq.C = length(cq)-1; 
Mq.D = length(dq)-1; 
Mq.T = T;
Mq.delay = 1;
Mq.type  = 'bj';

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Estimate on basis of noise corrupted data
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

Gq = est(Z,Mq,OPT); 

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Plot the results
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

if dsp,
	Gt.A = aq;
	Gt.B = bq;
	Gt.C = cq;
	Gt.D = dq;
	Gt.T = T;
	Gt.type = 'bj';
	Gt.disp.colour='b';
	Gt.disp.legend = 'True Response';
	Gq.disp.legend = [Gq.disp.legend, ' q operator'];
	showbode(Gt,Gq);
end

echo off;

if dm
 disp('  ')
 disp('---------------------------------------------------------------------')
 disp('  ')
 disp('You now have access to the MATLAB workspace so that you may examine')
 disp('the results of this simulation.  To return to the demos, type "dbcont"')
 disp(' ')
 keyboard; 
end;

File: src/uonidtoolbox/demo/_demo_bj.py

import uonidtoolbox as unit
import numpy as np
import scipy


def demo_bj(disp=1):

    OPT = unit.struct()
    OPT.dsp = disp

    # ====================================================
    # Specify Experiment Conditions
    # ====================================================
    T   = 1     # Sampling period in seconds
    N   = 500   # Number of Samples
    var = 1e-2  # Measurement Noise Variance


    # ====================================================
    # Specify true (linear) system
    # ====================================================
    den         = np.real(np.poly([-0.1,-1,-0.001-0.2j,-0.001+0.2j,-0.5]))
    num         = 10*den[-1]
    sysc        = scipy.signal.TransferFunction(num, den).to_ss()
    A,B,C,D,_   = scipy.signal.cont2discrete((sysc.A, sysc.B, sysc.C, sysc.D), T, method='zoh')
    bq,aq       = scipy.signal.ss2tf(A, B, C, D, 0)
    bq,aq       = np.squeeze(bq), np.squeeze(aq)
    cq          = np.array([1,-0.2])
    dq          = np.array([1,-0.5])


    # ====================================================
    # Simulate a data record
    # ====================================================
    Z       = unit.struct()
    t       = np.arange(0, N, 1)
    Z.u     = np.sign(np.sin(3*np.pi*t))
    noise   = np.sqrt(var)*np.random.randn(Z.u.size).reshape(Z.u.shape)
    noise   = scipy.signal.lfilter(cq, dq, noise)
    Z.y     = scipy.signal.lfilter(bq, aq, Z.u) + noise


    # ====================================================
    # Specify Model Structures
    # ====================================================
    Mq          = unit.struct()
    Mq.A        = aq.shape[0]-1
    Mq.B        = bq.shape[0]-2
    Mq.C        = cq.shape[0]-1
    Mq.D        = dq.shape[0]-1
    Mq.T        = T
    Mq.delay    = 1
    Mq.type     = 'bj'


    # ====================================================
    # Estimate on basis of noise corrupted data
    # ====================================================
    Gq = unit.est(Z,Mq,OPT)


    # ====================================================
    # Plot the results
    # ====================================================
    if OPT.dsp:
        Gt              = unit.struct()
        Gt.disp         = unit.struct()
        Gt.A            = aq
        Gt.B            = bq
        Gt.C            = cq
        Gt.D            = dq
        Gt.T            = T
        Gt.w            = Gq.w
        Gt.type         = 'bj'
        # Gt.colour       = 'b'
        Gt.disp.legend  = 'True Response'
        Gt.disp.legend  = Gt.disp.legend + ' q operator'
        
        unit.plotting.showbode([Gt, Gq])
    #endif

#endfunction

